src/main/java/se/liu/ida/hefquin/engine/data/mappings/impl/EntityMappingImpl.java

```diff
@@ -25,26 +30,152 @@ public EntityMappingImpl( final Graph mappingDescription ) {
+			if(tp.asJenaTriple().getObject().isURI()) {
+			final Set<Node> localObjects = g2lMap.get(tp.asJenaTriple().getObject());
+			if(g2lMap.containsKey(tp.asJenaTriple().getObject())) {
+				for (final Triple workingTriple : workingSet) {
+					for (final Node localObject : localObjects) {
+						final Triple resultTriple = new Triple(workingTriple.getSubject(), workingTriple.getPredicate(), localObject);
+						resultSet.add(new TriplePatternImpl(resultTriple));
+					}
+				}
+				return resultSet;
+			} else {
+				for (final Triple workingTriple : workingSet ) {
+					resultSet.add(new TriplePatternImpl(workingTriple));
+				}
+				return resultSet;
+			}
+		} else {
+			for (final Triple workingTriple : workingSet ) {
+				resultSet.add(new TriplePatternImpl(workingTriple));
+			}
+			return resultSet;
+		}
 	}
 
 	@Override
 	public Set<SolutionMapping> applyToSolutionMapping( final SolutionMapping sm ) {
-		// TODO: implement this function
-		return Collections.singleton(sm);
+		final Set<Map<Var,Node>> cartesianProduct = new HashSet<>();
+		final Binding binding = sm.asJenaBinding();
+		final Iterator<Var> it = binding.vars();
+		
+		final Var firstVar = it.next(); // Create a number of single var-node combinations to populate the set with.
+		final Node firstNode = binding.get(firstVar);
+		if (firstNode.isURI()) {
+			final Set<Node> mappedNodes = g2lMap.get(firstNode);
+			if (mappedNodes == null) { // Local Node not different from global. 
+				final Map<Var,Node> newMap = new HashMap<>();
+				newMap.put(firstVar,firstNode);
+				cartesianProduct.add(newMap);
+			} else { // Local different from global exists, use that.
+				for (Node node : mappedNodes) {
+					final Map<Var,Node> newMap = new HashMap<>();
+					newMap.put(firstVar,node);
+					cartesianProduct.add(newMap);
+				}
+			}
+		} else {
+			final Map<Var,Node> newMap = new HashMap<>();
+			newMap.put(firstVar,firstNode);
+			cartesianProduct.add(newMap);
+		}
+		
+		while (it.hasNext()) {
+			final Var var = it.next();
+			final Node node = binding.get(var);
+			if (node.isURI()) {
+				final Set<Node> mappedNodes = g2lMap.get(node);
+				if (mappedNodes == null) { // Local Node not different from global. 
+					for(Map<Var,Node> map : cartesianProduct) { // Multiply by one by adding the pair to all existing combinations.
+						map.put(var, node);
+					}
+				} else { // Local different from global exists, use that.
+					final Set<Map<Var,Node>> newCartesianProducts = new HashSet<>();
+					for(Map<Var,Node> map : cartesianProduct) {
+						final Iterator<Node> nodeIt = mappedNodes.iterator();
+						final Node newNode = nodeIt.next(); // Add the first of the combinations to each map.
+						map.put(var, newNode);
+						
+						while (nodeIt.hasNext()) { // Add the rest of the multiplications from 2 thru n, if such exists.
+							final Node anotherNewNode = it.next();
+							final Map<Var,Node> newMap = new HashMap<>();
+							newMap.putAll(map);
```

Comment:
Better do the following, which achieves.
```suggestion
							final Map<Var,Node> newMap = new HashMap<>(map);
```